rig-neo4j/src/lib.rs

```diff
@@ -0,0 +1,297 @@
+        // TODO: Implement
+        Ok(None)
+    }
+
+    async fn get_document_by_query(
+        &self,
+        _query: Self::Q,
+    ) -> Result<Option<DocumentEmbeddings>, VectorStoreError> {
+        // TODO: Implement
+        Ok(None)
+    }
+}
+
+impl Neo4jVectorStore {
+    pub fn new(client: Neo4jClient, database_name: &str) -> Self {
+        Self {
+            client,
+            database_name: database_name.to_string(),
+        }
+    }
+    /// Creates a `Neo4jVectorIndex` that mirrors an existing Neo4j Vector Index.
+    ///
+    /// An index (of type "vector") of the same name as `index_name` must already exist for the Neo4j database.
+    /// See the Neo4j [documentation (Create vector index)](https://neo4j.com/docs/genai/tutorials/embeddings-vector-indexes/setup/vector-index/) for more information on creating indexes.
+    ///
+    /// ❗IMPORTANT: The index must be created with the same embedding model that will be used to query the index.
+    pub fn index<M: EmbeddingModel>(
+        &self,
+        model: M,
+        index_name: &str,
+        search_params: SearchParams,
+    ) -> Neo4jVectorIndex<M> {
+        Neo4jVectorIndex::new(self.client.graph.clone(), model, index_name, search_params)
+    }
+}
+
+/// A vector index for a Neo4j graph.
+pub struct Neo4jVectorIndex<M: EmbeddingModel> {
+    //collection: mongodb::Collection<DocumentEmbeddings>,
+    graph: Graph,
+    embedding_model: M,
+    index_name: String,
+    search_params: SearchParams,
+}
+
+impl<M: EmbeddingModel> Neo4jVectorIndex<M> {
+    const BASE_VECTOR_SEARCH_QUERY: &str = "
+    CALL db.index.vector.queryNodes($index_name, $num_candidates, $queryVector)
+    YIELD node, score
+    ";
+
+    pub fn new(
+        graph: Graph,
+        embedding_model: M,
+        index_name: &str,
+        search_params: SearchParams,
+    ) -> Self {
+        Self {
+            graph,
+            embedding_model,
+            index_name: index_name.to_string(),
+            search_params,
+        }
+    }
+
+    /// Build a Neo4j query that performs a vector search against an index.
+    /// See [Query vector index](https://neo4j.com/docs/cypher-manual/current/indexes/semantic-indexes/vector-indexes/#query-vector-index) for more information.
+    pub fn build_vector_search_query(
+        &self,
+        prompt_embedding: Embedding,
+        return_node: bool,
+        n: usize,
+    ) -> Query {
+        let where_clause = match &self.search_params.post_vector_search_filter {
+            Some(filter) => format!("WHERE {}", filter),
+            None => "".to_string(),
+        };
+
+        Query::new(format!(
+            "
+            {}
+            {}
+            RETURN score, ID(node) as element_id {}
+            ",
+            Self::BASE_VECTOR_SEARCH_QUERY,
+            where_clause,
+            if return_node { ", node as node" } else { "" }
+        ))
+        .param("queryVector", prompt_embedding.vec)
```

Comment:
Is it intended that some params are camelCased while others are snake_cased?

Reply:
nope, thanks! These are queries from the neo4j docs